osmnx/geometries.py

```diff
@@ -823,17 +833,74 @@ def _subtract_inner_polygons_from_outer_polygons(element, outer_polygons, inner_
     return geometry
 
 
-def _filter_final_gdf(gdf, polygon, tags):
+def _buffer_invalid_geometries(GDF):
+    """
+    Buffer any invalid geometries remaining in the GeoDataFrame.
+
+    Invalid geometries in the GeoDataFrame (which may accurately
+    reproduce invalid geometries in OpenStreetMap) can cause the
+    filtering to the query polygon and other subsequent geometric
+    operations to fail. This function logs the ids of the invalid
+    geometries and applies a buffer of zero to try to make them valid.
+
+    Note: the resulting geometries may differ from the originals
+    - please check them against OpenStreetMap
+
+    Parameters
+    ----------
+    GDF : GeoDataFrame
+        the GeoDataFrame with potential invalid geometries
+
+    Returns
+    -------
+    GDF : GeoDataFrame
+        the GeoDataFrame with .buffer(0) applied to invalid geometries
+    """
+    # only apply the filters if the GeoDataFrame is not empty
+    if not GDF.empty:
+
+        # create a filter for rows with invalid geometries
+        invalid_geometry_filter = ~GDF["geometry"].is_valid
+
+        # if there are invalid geometries
+        if sum(invalid_geometry_filter) > 0:
+
+            # get their unique_ids from the index
+            invalid_geometry_ids = GDF.loc[invalid_geometry_filter].index.to_list()
+
+            # show a warning
+            warnings.warn(
+                "Some invalid geometries were created and included in the GeoDataFrame. "
+                "A buffer of 0 has been applied to try to make them valid "
+                f"and their unique ids have been logged. {invalid_geometry_ids}",
+                stacklevel=2,
+            )
+
+            # create a list of their urls and log them
+            osm_url = "https://www.openstreetmap.org/"
```

Comment:
I think just supplying the OSMID itself (rather than a full URL to it) is sufficient for users who want to look it up?

Reply:
Yes. It is partly a convenience at the moment as it creates a clickable link in the terminal log which takes you directly to that geometry on osm.org. If it's all right I'd like to keep it at the moment and can take it out once all of the testing is out of the way.

Reply:
Certainly.